### ISSUE TITLE
Widget gets built twice

### ISSUE THREAD

**Issue by GeorgePagounis** [flutter/flutter#86315]
## Steps to Reproduce

```int times = 0;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
      //routes: routes,
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    times++;
    print('HomePage is built $times times');
    return Scaffold(
      appBar: AppBar(),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

I/flutter (11413): HomePage is built 1 times

**Actual results:** I/flutter (11413): HomePage is built 1 times
                            I/flutter (11413): HomePage is built 2 times

Default flutter create app/ HomePage widget gets called twice during build. This is actually a serious issue. What if I want to initialize Hive in the first Widget of the app and show a splash screen while I do that? The future function would be called twice and exceptions would be thrown. Also duplicate of https://github.com/flutter/flutter/issues/33566 . I just opened another since that one is closed for some reason.

run flutter doctor -v):
[√] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.19042.1083], locale en-US)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.
[√] Chrome - develop for the web
[√] Android Studio (version 4.1.0)
[√] VS Code (version 1.58.0)
[√] Connected device (3 available)

**Comment by stuartmorgan** [comment]
> This is actually a serious issue. What if I want to initialize Hive in the first Widget of the app and show a splash screen while I do that?

If you are trying to use `build` for doing one-time initialization, you are not using `State` correctly. You should read [the documentation for `State`](https://api.flutter.dev/flutter/widgets/State-class.html) which:
- explains where to do one-time initialization
- explicitly says that `build` may be called any number of times.

Can you describe what actual issue this is causing?

**Comment by GeorgePagounis** [comment]
Ok, so since a widget statefull or not, can be randomly rebuilt when not implicitly programmed or notified (setState and other notifyListeners methods) from the programmer, instead of this code:

```import 'package:flutter/material.dart';

final Map<String, WidgetBuilder> routes = <String, WidgetBuilder>{
  '/': (BuildContext context) {
    print('This route is executed $times times');
    return MyHomePage();
  },
};

int times = 0;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      //home: MyHomePage(),
      routes: routes,
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    times++;
    print('HomePage is built $times times');

    Future<void> wait() async {
      print('one time init is happening $times times');
      await Future.delayed(Duration(seconds: 2));
    }

    return FutureBuilder(
        future: wait(),
        builder: (context, snapshot) {
          return snapshot.connectionState == ConnectionState.done
              ?
              //what should appear after one time init
              Scaffold(
                  body: Center(
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        Text(
                          'You have pushed the button this many times:',
                        ),
                        Text(
                          '',
                          style: Theme.of(context).textTheme.headline4,
                        ),
                      ],
                    ),
                  ),
                  floatingActionButton: FloatingActionButton(
                    onPressed: () {},
                    tooltip: 'Increment',
                    child: Icon(Icons.add),
                  ),
                )
              : Scaffold(
                  body: CircularProgressIndicator(),
                ); //SplashScreen
        });
  }
}
```
MyHomePage Widget being Statefull or Stateless doesn't matter at this point

I should do it like this:

```
import 'package:flutter/material.dart';

final Map<String, WidgetBuilder> routes = <String, WidgetBuilder>{
  '/': (BuildContext context) {
    print('This route is executed $times times');
    return MyHomePage();
  },
};

int times = 0;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      //home: MyHomePage(),
      routes: routes,
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool done = false;

  @override
  void initState() {
    wait();
    super.initState();
  }

  Future<void> wait() async {
    print('one time init is happening $times times');
    await Future.delayed(Duration(seconds: 2));
    setState(() {
      done = true;
    });
  }

  @override
  Widget build(BuildContext context) {
    times++;
    print('HomePage is built $times times');

    return done
        ?
        //what should appear after one time init
        Scaffold(
            body: Center(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text(
                    'You have pushed the button this many times:',
                  ),
                  Text(
                    '',
                    style: Theme.of(context).textTheme.headline4,
                  ),
                ],
              ),
            ),
            floatingActionButton: FloatingActionButton(
              onPressed: () {},
              tooltip: 'Increment',
              child: Icon(Icons.add),
            ),
          )
        : Scaffold(
            body: CircularProgressIndicator(),
          ); //SplashScreen
  }
}
```

Doesn't this however make the FutureBuilder Widget useless, what if I wanted to read a document from Firestore, or any backend for that matter, wouldn't that make unnecessary reads? Also keep in mind that the output with HomePage being on routes is:

I/flutter (12195): This route is executed 0 times
I/flutter (12195): one time init is happening 0 times
I/flutter (12195): HomePage is built 1 times
I/flutter (12195): This route is executed 1 times
I/flutter (12195): HomePage is built 2 times
I/flutter (12195): HomePage is built 3 times


Is this also a wanted behaviour? Why would the navigator be called twice and if I had some functions that lived there, it could have certain implications.

**Comment by TahaTesser** [comment]
Hi @GeorgePagounis 
This looks more like a logic issue than a Flutter bug,  Please see https://flutter.dev/community for resources and asking questions like this, you may also get some help if you post it on Stack Overflow and if you need help with your code, please see https://www.reddit.com/r/flutterhelp/
Closing, as this isn't an issue with Flutter itself. If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by yakeer** [comment]
Happens to me as well,
so in order to eliminate the option that it's a logic problem in my code,
I created a new flutter project, and ran it out of the box.

The build has hit twice.
Using flutter 2.2.3. Happens also on MacOS and also on Windows. using Visual Studio code.

**Comment by stuartmorgan** [comment]
> in order to eliminate the option that it's a logic problem in my code

The comment above wasn't saying that there's a logic error cause build to be called multiple times; the logic error would be writing code assuming that build can only be called once.

**Comment by yakeer** [comment]
> > in order to eliminate the option that it's a logic problem in my code
> 
> The comment above wasn't saying that there's a logic error cause build to be called multiple times; the logic error would be writing code assuming that build can only be called once.

But in general speaking,
we assume that a widget build function happens only once if nothing changes in its subtree.

Is this a wrong assumption?
